File: backend/agent/tool_path.go
package agent

import (
	"fmt"
	"path/filepath"
	"strings"
)

func (a *Agent) resolvePath(p string) (string, error) {
	if p == "" {
		return "", fmt.Errorf("path is required")
	}

	var abs string
	if filepath.IsAbs(p) {
		abs = filepath.Clean(p)
	} else {
		abs = filepath.Clean(filepath.Join(a.workDir, p))
	}

	workAbs, err := filepath.Abs(a.workDir)
	if err != nil {
		return "", err
	}

	if abs != workAbs && !strings.HasPrefix(abs, workAbs+string(filepath.Separator)) {
		return "", fmt.Errorf("path escapes workspace: %s", p)
	}
	return abs, nil
}

File: backend/agent/tool_path.ts
import { Effect } from "effect";
import path from "node:path";
import { Agent } from "./agent";

Agent.prototype.resolvePath = function (this: Agent, p: string): Effect.Effect<string, Error> {
  if (p === "") {
    return Effect.fail(new Error("path is required"));
  }

  let abs: string;
  if (path.isAbsolute(p)) {
    abs = path.normalize(p);
  } else {
    abs = path.normalize(path.join(this.workDir, p));
  }

  return Effect.try({
    try: () => {
      return path.resolve(abs);
    },
    catch: (cause) => (cause instanceof Error ? cause : new Error(String(cause))),
  });
};

